started versioning (#2407)
conan-io/docs: tutorial/consuming_packages/intro_to_versioning.rst

Version ranges
--------------

A ``requires`` can express a dependency to a certain range of versions for a given package, with the syntax ``pkgname/[version-range-expression]``.
Lets see an example, please, first clone the sources to recreate this project. You can find them in the
`examples2.0 repository <https://github.com/conan-io/examples2>`_ in GitHub:

.. code-block:: bash

    $ git clone https://github.com/conan-io/examples2.git
    $ cd examples2/tutorial/consuming_packages/versioning

We can see that we have there:

.. code-block:: python
    :caption: **conanfile.py**

    from conan import ConanFile

    class CompressorRecipe(ConanFile):
        settings = "os", "compiler", "build_type", "arch"
        generators = "CMakeToolchain", "CMakeDeps"

        def requirements(self):
            self.requires("zlib/[~1.2]")

That ``requires`` contains the expression ``zlib/[~1.2]``, which means "approximately" ``1.2`` version, that means, it can resolve to
any ``zlib/1.2.8``, ``zlib/1.2.11`` or ``zlib/1.2.12``, but it will not resolve to something like ``zlib/1.3.0``. Among the available
matching versions, a version range will always pick the latest one.

> If we do a conan install, we would see something like:

Reviewer: ```suggestion
If we do a :command:`conan install`, we would see something like:
```

Then we would see it resolved to the latest available CMake package, with at least version ``3.11``:

.. code-block:: bash

    $ conan install .
    ...
    Graph root
        conanfile.py: .../conanfile.py
    Requirements
        zlib/1.2.12#87a7211557b6690ef5bf7fc599dd8349 - Cache
    Build requirements
        cmake/3.19.8#f305019023c2db74d1001c5afa5cf362 - Downloaded
    Resolved version ranges
        cmake/[>3.10]: cmake/3.19.8
        zlib/[~1.2]: zlib/1.2.12

Revisions
---------

What happens when a package creator does some change to the package recipe or to the source code, but they don't bump the ``version`` 
to reflect those changes? Conan has an internal mechanism to keep track of those modifications, and it is called the **revisions**.

The recipe revision is the hash that can be seen together with the package name and version in the form ``pkgname/version#recipe_revision``
or ``pkgname/version@user/channel#recipe_revision``.
The recipe revision is a hash of the contents of the recipe and the source code. So if something changes either in the recipe,
its associated files or in the source code that this recipe is packaging, it will create a new recipe revision.

> You can list existing revisions with the ``conan list`` command:

Reviewer: ```suggestion
You can list existing revisions with the :command:`conan list` command:
```

Such snapshot must be "realizable", that is, it needs to be a state that can be actually reproduce from the conanfile recipes.
And this lockfile can be used at a later point in time to force that same state, even if there are new created package versions.

Lets see lockfiles in action. First, lets pin the dependency to ``zlib/1.2.11`` in our example:

.. code-block:: python

    def requirements(self):
        self.requires("zlib/1.2.11")

And lets capture a lockfile:

.. code-block:: bash

    conan lock create .

    -------- Computing dependency graph ----------
    Graph root
        conanfile.py: .../conanfile.py
    Requirements
        zlib/1.2.11#4524fcdd41f33e8df88ece6e755a5dcc - Cache

    Generated lockfile: .../conan.lock

Lets see what the lockfile ``conan.lock`` contains:

.. code-block:: json

    {
        "version": "0.5",
        "requires": [
            "zlib/1.2.11#4524fcdd41f33e8df88ece6e755a5dcc%1650538915.154"
        ],
        "build_requires": [],
        "python_requires": []
    }

Now, lets restore the original ``requires`` version range:

.. code-block:: python

    def requirements(self):
        self.requires("zlib/[~1.2]")

> And run ``conan install .``, which by default will find the ``conan.lock``, and run the equivalent ``conan install . --lockfile=conan.lock``
> 

Reviewer: ```suggestion
And run :command:`conan install .`, which by default will find the ``conan.lock``, and run the equivalent :command:`conan install . --lockfile=conan.lock`

```

.. code-block:: bash

    conan install .

    Graph root
        conanfile.py: .../conanfile.py
    Requirements
        zlib/1.2.11#4524fcdd41f33e8df88ece6e755a5dcc - Cache

Note how the version range is no longer resolved, and it doesn't get the ``zlib/1.2.12`` dependency, even if it is the 
allowed range ``zlib/[~1.2]``, because the ``conan.lock`` lockfile is forcing it to stay in ``zlib/1.2.11`` and that exact revision too.

Read more
---------

Go to :ref:`this section <tutorial_versioning_ci>` to learn more.

> 

Reviewer: .. container:: examples

    - :ref:`Introduction to Versioning and Continuous Integration <tutorial_versioning_ci>`

> Go to :ref:`this section <tutorial_versioning_ci>` to learn more.
> 

Reviewer: ```suggestion
.. container:: examples

    - :ref:`Introduction to Versioning and Continuous Integration <tutorial_versioning_ci>`

```

Author: But these are not examples, but another section in the docs. The container is a "ReadMore" one or an "Examples" one?

Reviewer: I thought we were not differentiating between those... then we can leave it like this?

```suggestion
- :ref:`Introduction to Versioning and Continuous Integration <tutorial_versioning_ci>`

```
